modules/enterprise/server/jar/src/main/java/org/rhq/enterprise/server/storage/StorageClientManager.java

```diff
@@ -236,26 +240,36 @@ public synchronized boolean refreshCredentialsAndSession() {
 
         if ((username != null && !username.equals(this.cachedStorageUsername))
             || (password != null && !password.equals(this.cachedStoragePassword))) {
+            return refreshSession();
+        }
 
-            Session wrappedSession;
-            try {
-                wrappedSession = createSession();
-            } catch (NoHostAvailableException e) {
-                if (cluster != null) {
-                    cluster.shutdown();
-                }
+        return true;
+    }
 
-                LOG.warn("Storage client subsystem wasn't initialized because it wasn't possible to connect to the"
-                    + " storage cluster. The RHQ server is set to MAINTENANCE mode. Please start the storage cluster"
-                    + " as soon as possible.", e);
-                return false;
+    /**
+     * Recreates the session used to connect to the storage node.
+     * @return true if success, false otherwise.
+     */
+    private synchronized boolean refreshSession() {
+        Session wrappedSession;
+        try {
+            wrappedSession = createSession();
```

Comment:
We should shut down the old session; otherwise, we will be leaking connections and threads.